### PR TITLE
TF-1808 store n recently opened email in cache

### PR DESCRIPTION
**Pull request by nqhhdev**
- issues: #1808 

- Resolved

`IOS`

- First time

https://github.com/linagora/tmail-flutter/assets/99852347/a2773fb2-6d64-483c-8093-6a0313a7c6f0

- Secondary time (Duplicate)

https://github.com/linagora/tmail-flutter/assets/99852347/65d324d3-4177-457e-a81d-9fae36de2c9f

`Android`

- First time

https://github.com/linagora/tmail-flutter/assets/99852347/7429dbad-529a-4271-b955-7206703bb493

- Secondary time (Duplicate)

https://github.com/linagora/tmail-flutter/assets/99852347/f342bebb-1dc2-48c9-babe-a4a32a0a509a

### CODE REVIEW

File: core/lib/utils/file_utils.dart [hunk 1]
```diff
@@ -73,6 +73,20 @@ class FileUtils {
 
     return emailContent;
   }
+
+  Future<bool> getPathFileExists({
```
Comment (hoangdat) [inline core/lib/utils/file_utils.dart:77]:
```suggestion
  Future<bool> isFileExisted({
```

File: lib/features/email/data/datasource/email_datasource.dart [hunk 2]
```diff
@@ -79,4 +79,6 @@ abstract class EmailDataSource {
   Future<Email> getDetailedEmailById(Session session, AccountId accountId, EmailId emailId);
 
   Future<void> storeEmail(Session session, AccountId accountId, Email email);
+
+  Future<void> storeOpenedEmailToCache(Session session, AccountId accountId, DetailedEmail detailedEmail);
```
Comment (hoangdat) [inline lib/features/email/data/datasource/email_datasource.dart:83]:
```suggestion
  Future<void> storeOpenedEmail(Session session, AccountId accountId, DetailedEmail detailedEmail);
```

File: lib/features/email/data/datasource_impl/email_datasource_impl.dart [hunk 3]
```diff
@@ -166,4 +166,9 @@ class EmailDataSourceImpl extends EmailDataSource {
   Future<void> storeEmail(Session session, AccountId accountId, Email email) {
     throw UnimplementedError();
   }
+
+  @override
+  Future<void> storeOpenedEmailToCache(Session session, AccountId accountId, DetailedEmail detailedEmail) {
```
Comment (hoangdat) [inline lib/features/email/data/datasource_impl/email_datasource_impl.dart:171]:
```suggestion
  Future<void> storeOpenedEmail(Session session, AccountId accountId, DetailedEmail detailedEmail) {
```

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart [hunk 4]
```diff
@@ -149,4 +152,37 @@ class EmailHiveCacheDataSourceImpl extends EmailDataSource {
...
     }).catchError(_exceptionThrower.throwException);
   }
+
+  @override
+  Future<void> storeOpenedEmailToCache(Session session, AccountId accountId, DetailedEmail detailedEmail) {
+    return Future.sync(() async {
+      final task = HiveTask(
+          runnable: () async {
+
+            final isDuplicateDetailedEmail = await _detailedEmailCacheManager.checkDuplicateOpenedEmail(accountId, session.username,detailedEmail);
```
Comment (hoangdat) [inline lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:162]:
```suggestion
            final detailedEmailExisted = await _detailedEmailCacheManager.checkDuplicateOpenedEmail(accountId, session.username,detailedEmail);
```

Comment (hoangdat) [reply lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:162]: And I think we need to change the name `checkDuplicateOpenedEmail` also.
How about `openedDetailsEmailCacheHit`

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart [hunk 5]
```diff
@@ -149,4 +152,37 @@ class EmailHiveCacheDataSourceImpl extends EmailDataSource {
...
+            log('EmailHiveCacheDataSourceImpl::storeOpenedEmailToCache():fileSavedPath: ${fileSaved.path}');
+
+            final detailedEmailSaved = detailedEmail.fromEmailContentPath(fileSaved.path);
+
+            await _detailedEmailCacheManager.storeOpenedEmail(
+              accountId,
+              session.username,
+              detailedEmailSaved);
+          });
+      return await _openedEmailCacheWorkerQueue.addTask(task);
```
Comment (hoangdat) [inline lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:185]: I think we dont need to `await` for `addTask`. WDYT?
cc: @dab246

File: lib/features/email/domain/repository/email_repository.dart [hunk 6]
```diff
@@ -93,4 +93,6 @@ abstract class EmailRepository {
   Future<Email> getDetailedEmailById(Session session, AccountId accountId, EmailId emailId);
 
   Future<void> storeEmailToCache(Session session, AccountId accountId, Email email);
+
+  Future<void> storeOpenedEmailToCache(Session session, AccountId accountId, DetailedEmail detailedEmail);
```
Comment (hoangdat) [inline lib/features/email/domain/repository/email_repository.dart:97]:
```suggestion
  Future<void> storeOpenedEmail(Session session, AccountId accountId, DetailedEmail detailedEmail);
```

**Comment by hoangdat** [comment]
Dont forget to clean the app's data when logout! @nqhhdev @dab246 Did we have the ticket for it?

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart [hunk 7]
```diff
@@ -149,4 +152,37 @@ class EmailHiveCacheDataSourceImpl extends EmailDataSource {
...
+
+            if (isDuplicateDetailedEmail) {
+              return Future.value();
+            }
+
+            final fileSaved = await _fileUtils.saveToFile(
+              nameFile: detailedEmail.emailId.asString,
+              content: detailedEmail.htmlEmailContent ?? '',
+              folderPath: detailedEmail.folderPath,
+              extensionFile: ExtensionType.text.value
```
Comment (dab246) [inline lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:173]: remove `ExtensionType.text.value`

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart (hunk 5 above)
Comment (dab246) [reply lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:185]: Should remove `await`

File: lib/features/email/data/repository/email_repository_impl.dart [hunk 8]
```diff
@@ -198,4 +198,9 @@ class EmailRepositoryImpl extends EmailRepository {
   Future<void> storeEmailToCache(Session session, AccountId accountId, Email email) {
     return emailDataSource[DataSourceType.hiveCache]!.storeEmail(session, accountId, email);
   }
+
+  @override
+  Future<void> storeOpenedEmailToCache(Session session, AccountId accountId, DetailedEmail detailedEmail) {
```
Comment (dab246) [inline lib/features/email/data/repository/email_repository_impl.dart:203]: rename to `storeOpenedEmailToCache` to `storeOpenedEmail`

File: lib/features/email/domain/state/store_opened_email_to_cache_state.dart [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import 'package:core/core.dart';
+
+class StoreOpenedEmailToCacheLoading extends UIState {}
+
+class StoreOpenedEmailToCacheSuccess extends UIState {}
```
Comment (dab246) [inline lib/features/email/domain/state/store_opened_email_to_cache_state.dart:5]: Rename to `StoreOpenedEmailSuccess `

File: lib/features/offline_mode/manager/detailed_email_cache_manager.dart [hunk 10]
```diff
@@ -58,4 +63,59 @@ class DetailedEmailCacheManager {
...
+
+  Future<bool> _getDetailedEmailCache(
+    AccountId accountId,
+    UserName userName,
+    DetailedEmail detailedEmail
+  ) async {
+    final keyCache = TupleKey(detailedEmail.emailId.asString, accountId.asString, userName.value).encodeKey;
+    final detailedEmailCache = await _cacheClient.getItem(keyCache);
+    log('DetailedEmailCacheManager::_getDetailedEmailCache():_getDetailedEmailCache: $detailedEmailCache');
+    if (detailedEmailCache != null) {
```
Comment (dab246) [inline lib/features/offline_mode/manager/detailed_email_cache_manager.dart:115]: Should write `return detailedEmailCache != null`

File: model/lib/email/presentation_email.dart [hunk 11]
```diff
@@ -31,6 +32,7 @@ class PresentationEmail with EquatableMixin {
   final SelectMode selectMode;
   final Uri? routeWeb;
   final PresentationMailbox? mailboxContain;
+  final List<EmailHeader>? emailHeader;
```
Comment (dab246) [inline model/lib/email/presentation_email.dart:35]: check in emailExtension

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  final OpenedEmailHiveCacheClient _cacheClient;
+  final FileUtils _fileUtils;
+
+  OpenedEmailCacheManager(this._cacheClient, this._fileUtils);
+
+  Future<void> insertDetailedEmail(
+      AccountId accountId,
+      UserName userName,
+      DetailedEmailHiveCache detailedEmailCache
+      ) {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:26]: remove indent

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    detailedEmailCacheList.sortByLatestTime();
+    log('OpenedEmailCacheManager::getAllDetailedEmails():SIZE: ${detailedEmailCacheList.length}');
+    return detailedEmailCacheList;
+  }
+
+  Future<void> storeOpenedEmail(
+      AccountId accountId,
+      UserName userName,
+      DetailedEmail detailedEmail
+      ) async {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:53]: same

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+      await removeDetailedEmail(accountId, userName, latestEmail.emailId);
+    }
+    await insertDetailedEmail(accountId, userName, detailedEmail.toHiveCache());
+  }
+
+  Future<bool> openedDetailsEmailCacheHit(
+      AccountId accountId,
+      UserName userName,
+      DetailedEmail detailedEmail
+      ) async {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:68]: same

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    final emailContentPathExists = await _getEmailContentPath(detailedEmail);
+    final detailedEmailCacheExists = await _getDetailedEmailCache(accountId, userName, detailedEmail);
+
+    if (emailContentPathExists == true && detailedEmailCacheExists) {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:72]: should write `return emailContentPathExists == true && detailedEmailCacheExists`

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  Future<bool?> _getEmailContentPath(DetailedEmail detailedEmail) async {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:79]: rename to `isFileExisted`

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    );
+    log('OpenedEmailCacheManager::_getDetailedEmailCache():_getEmailContentPath: $fileSaved');
+    return fileSaved;
+  }
+
+  Future<bool> _getDetailedEmailCache(
+      AccountId accountId,
+      UserName userName,
+      DetailedEmail detailedEmail
+      ) async {
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:92]: remove indent

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart [hunk 18]
```diff
@@ -149,4 +154,36 @@ class EmailHiveCacheDataSourceImpl extends EmailDataSource {
...
+              return Future.value();
+            }
+
+            final fileSaved = await _fileUtils.saveToFile(
+              nameFile: detailedEmail.emailId.asString,
+              content: detailedEmail.htmlEmailContent ?? '',
+              folderPath: detailedEmail.folderPath
+            );
+
+            log('EmailHiveCacheDataSourceImpl::storeOpenedEmailToCache():fileSavedPath: ${fileSaved.path}');
```
Comment (dab246) [inline lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:177]: rename to `log('EmailHiveCacheDataSourceImpl::storeOpenedEmail():fileSavedPath: ${fileSaved.path}');`

File: lib/features/email/domain/state/store_opened_email_to_cache_state.dart [hunk 19]
```diff
@@ -0,0 +1,13 @@
+import 'package:core/core.dart';
```
Comment (dab246) [inline lib/features/email/domain/state/store_opened_email_to_cache_state.dart:1]: Import only files that are used

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+
+    if (listDetailedEmails.length >= CachingConstants.maxNumberOpenedEmailsForOffline) {
+      final latestEmail = listDetailedEmails.last;
+      log('OpenedEmailCacheManager::handleStoreDetailedEmail():latestEmail: $latestEmail');
+      await removeDetailedEmail(accountId, userName, latestEmail.emailId);
+    }
+    await insertDetailedEmail(accountId, userName, detailedEmail.toHiveCache());
+  }
+
+  Future<bool> openedDetailsEmailCacheHit(
```
Comment (dab246) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:64]: rename to `isOpenedDetailEmailCached()`

**Comment by dab246** [comment]
> Dont forget to clean the app's data when logout! @nqhhdev @dab246 Did we have the ticket for it?

Don't worry, we already have a ticket for this issue. #1838

File: lib/features/offline_mode/manager/opened_email_cache_manager.dart [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+  Future<bool?> _isFileExisted(DetailedEmail detailedEmail) async {
+    final fileSaved = await _fileUtils.isFileExisted(
+      nameFile: detailedEmail.emailId.asString,
+      folderPath: detailedEmail.folderPath,
+    );
+    log('OpenedEmailCacheManager::_getDetailedEmailCache():_getEmailContentPath: $fileSaved');
+    return fileSaved;
+  }
+
+  Future<bool> _getDetailedEmailCache(
```
Comment (hoangdat) [inline lib/features/offline_mode/manager/opened_email_cache_manager.dart:84]: Why a get function return a bool value?

Comment (nqhhdev) [reply lib/features/offline_mode/manager/opened_email_cache_manager.dart]: This func is used to check if the email is cached or not.

I'll change func name `_isDetailEmailExistedInCache`, it's clearly

**Review by hoangdat — APPROVED** [review]

**Review by dab246 — APPROVED** [review]

**Pull request merged**